src/sys/windows/selector.rs

```diff
@@ -371,14 +411,233 @@ impl Selector {
+        /// for the user of the registered `Source`.
+        pub fn register_handle<H>(&self, handle: &H) -> io::Result<()> where H: AsRawHandle + ?Sized {
+            let handler = RawHandleCompletionHandler {
+                token: self.token
+            };
+            self.selector.cp.register_handle(handle, handler.into())
+        }
+
+        /// Inject an event to the event loop.
+        ///
+        /// In case it is required to generate an event even though there was no callback due to a
+        /// completed IO operation, this method can be used to mark the associated token as ready, e.g.
+        /// in case it is necessary to signal an initial readiness before any IO operation has been
+        /// scheduled.
+        pub fn inject_event(&self, readiness: Readiness) {
+            let event = readiness.into_event(self.token);
+            self.selector.injected_events.lock().unwrap().push(event)
+        }
+
+        /// Check whether the provided registry is the same as the one used during creation of the
+        /// binding.
+        ///
+        /// IOCP driven sources cannot unregister from their completion port before they get closed.
+        /// Therefore, implementations of reregister and deregister should check whether they have been
+        /// called on the same `Registry` using this method to report errors in case this constraint
+        /// has been violated.
+        pub fn is_same_registry(&self, registry: &Registry) -> bool {
+            Arc::ptr_eq(&self.selector, &poll::selector(registry).inner)
+        }
+    }
+
+    /// A wrapper around an internal instance over `miow::Overlapped` which is in
+    /// turn a wrapper around the Windows type `OVERLAPPED`.
+    ///
+    /// This type is required to be used for all IOCP operations on handles that are
+    /// registered with an event loop. The event loop will receive notifications
+    /// over `OVERLAPPED` pointers that have completed, and it will cast that
+    /// pointer to a pointer to this structure and invoke the associated callback.
+    #[repr(C)]
+    pub struct Overlapped {
+        inner: UnsafeCell<miow::Overlapped>,
+        #[cfg(debug_assertions)]
+        canary: u32,
+        callback: Box<dyn CompletionCallback>,
+    }
+
+    impl fmt::Debug for Overlapped {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("Overlapped").finish()
+        }
+    }
+
+    impl Overlapped {
+        /// Creates a new `Overlapped` which will invoke the provided `cb` callback
+        /// whenever it's triggered.
+        ///
+        /// The returned `Overlapped` must be used as the `OVERLAPPED` passed to all
+        /// I/O operations that are registered with mio's event loop. When the I/O
+        /// operation associated with an `OVERLAPPED` pointer completes the event
+        /// loop will invoke the object provided as `callback`.
+        pub fn new<C: CompletionCallback+'static>(callback: C) -> Self {
```

Comment:
Why do we need a trait with a single method? Can't we use`C: FnMut(&OVERLAPPED_ENTRY) + static`?

Reply:
Also can't Mio provide that implementation? Or is there special handling per type of handle?

Reply:
I was torn on this too, but I've chosen to introduce a trait because it can be extended in a source-compatible way by adding more trait methods with default impls if we find more cases of hooks that could be costomized by mio drivers (e.g. a "pre-poll hook").

There is a blanket impl for FnMuts as well, so they can be used, see the provided sample test case, which is doing exactly that.